sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -131,8 +131,20 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with PredicateHelper {
     expressions.flatMap(collectEvaluableUDFs)
   }
 
-  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case plan: LogicalPlan => extract(plan)
+  def apply(plan: LogicalPlan): LogicalPlan = plan match {
+    // SPARK-26293: A subquery will be rewritten into join later, and will go through this rule
+    // eventually. Here we skip subquery, as Python UDF only needs to be extracted once.
+    case _: Subquery => plan
```

Comment:
Personally I found it a bit confusing when two seeming unrelated things are put together (Subquery and ExtractPythonUDFs).

I wonder if it's sufficient to make ExtractPythonUDFs idempotent?

Reply:
I see. If it's common to skip Subquery in other rules, I guess it's ok to put it in here as well. But it would definitely be helpful to establish some kind of guidance, maybe sth like "All optimizer rule should skip Subquery because OptimizeSubqueries will execute them anyway"?

Reply:
I think you have a point here. If subquery will be converted to join, why do we need to optimize subquery ahead?

Anyway, that's something we need to discuss later. cc @dilipbiswal for the subquery question.

Reply:
I'm not sure if it is totally ok to skip `Subquery` for all optimizer rules.

For `ExtractPythonUDFs` I think it is ok because `ExtractPythonUDFs` is performed after the rules in `RewriteSubquery`. So we can skip `ExtractPythonUDFs` here and extract Python UDF after the subqueries are rewritten into join.

But for the rules which perform before `RewriteSubquery`, if we skip it on `Subquery`, we have no chance to do the rules after the subqueries are rewritten into join.

Reply:
Basically, we want to ensure this rule is running once and only once. In the future, if we have another rule/function that calls Optimizer.this.execute(plan), this rule needs to be fixed again... We have a very strong hidden assumption in the implementation. This looks risky in the long term.

The current fix is fine for backporting to 2.4.